addons/web/controllers/binary.py

```diff
@@ -84,30 +84,58 @@ def content_common(self, xmlid=None, model='ir.attachment', id=None, field='raw'
-        '/web/assets/<int:id>-<string:unique>/<path:extra>/<string:filename>'], type='http', auth="public")
+    @http.route([
+        '/web/assets/<path:extra>/<string:filename>',
+        '/web/assets/<string:unique>/<path:extra>/<string:filename>'], type='http', auth="public")
     # pylint: disable=redefined-builtin,invalid-name
-    def content_assets(self, id=None, filename=None, unique=False, extra=None, nocache=False):
-        if not id:
-            domain = [('url', '!=', False)]
-            if extra:
-                domain += [('url', '=like', f'/web/assets/%/{extra}/{filename}')]
-            else:
-                domain += [
-                    ('url', '=like', f'/web/assets/%/{filename}'),
-                    ('url', 'not like', f'/web/assets/%/%/{filename}')
-                ]
+    def content_assets(self, filename=None, unique='%', extra='-', nocache=False):
+        debug_assets = unique == 'debug'
+        attachment = None
+        if unique != 'debug':
+            url = f'/web/assets/{unique}/{extra}/{filename}'
+            domain = [
+                ('public', '=', True),
+                ('url', '!=', False),
+                ('url', '=like', url),
+                ('res_model', '=', 'ir.ui.view'),
+                ('res_id', '=', 0),
+            ]
             attachment = request.env['ir.attachment'].sudo().search(domain, limit=1)
-            if not attachment:
+        if not attachment:
+            # try to generate one
+            extra_parts = extra.split('_')
+            try:
+                params, rtl = request.env['ir.asset']._parse_assets_extra(extra_parts)
+            except ValueError:
+                _logger.exception('Invalid assets')
                 raise request.not_found()
-            id = attachment.id
+            if debug_assets:
+                bundle_name, asset_type = filename.rsplit('.', 1)
+            else:
+                bundle_name, _min, asset_type = filename.rsplit('.', 2)
+                assert _min == 'min'
+            css = asset_type == 'css'
+            js = asset_type == 'js'
+            assert js or css
+            bundle = request.env['ir.qweb']._get_asset_bundle(
+                bundle_name,
+                css=css,
+                js=js,
+                debug_assets=debug_assets,
+                rtl=rtl,
+                assets_params=params
+            )
+            # check if the version matches. If not, redirect to the last version
+            if not debug_assets and unique != '%' and unique != bundle.get_version(asset_type):
+                return request.redirect(bundle.get_link(asset_type), code=302)
+            if css and bundle.stylesheets:
+                attachment = bundle.css()
+            elif js and bundle.javascripts:
+                attachment = bundle.js()
+        if not attachment:
+            raise request.not_found()
         with replace_exceptions(UserError, by=request.not_found()):
```

Comment:
This reraise as 404 was primarily present to hide AccessErrors in case the given attachment id wasn't an asset bundle. With your changes you give the guarantee that it is an asset bundle, I think we can remove this line